Source/WebCore/page/SecurityOriginData.cpp

```diff
@@ -176,18 +176,17 @@ bool SecurityOriginData::shouldTreatAsOpaqueOrigin(const URL& url)
     if (!url.isValid())
         return true;
 
-    // FIXME: Do we need to unwrap the URL further?
-    URL innerURL = SecurityOrigin::shouldUseInnerURL(url) ? SecurityOrigin::extractInnerURL(url) : url;
-    if (!innerURL.isValid())
+    URL extractedURL = url.protocolIsBlob() ? SecurityOrigin::extractURLFromBlobURLPath(url) : url;
+    if (!extractedURL.isValid() || (url.protocolIsBlob() && !extractedURL.protocolIsInHTTPFamily()))
         return true;
 
     // For edge case URLs that were probably misparsed, make sure that the origin is opaque.
     // This is an additional safety net against bugs in URL parsing, and for network back-ends that parse URLs differently,
     // and could misinterpret another component for hostname.
-    if (schemeRequiresHost(innerURL) && innerURL.host().isEmpty())
+    if (schemeRequiresHost(extractedURL) && extractedURL.host().isEmpty())
         return true;
```

Comment:
I think we can turn this into an assert with our new URL implementation. Happy to make that change if you agree or attempt as a follow-up.